JPS_BASE_LIB/src/main/java/uk/ac/cam/cares/jps/base/derivation/DerivationSparql.java

```diff
@@ -223,21 +236,40 @@ List<String> unifiedBulkCreateDerivations(List<List<String>> entitiesList, List<
 			throw new JPSRuntimeException(errmsg);
 		}
 
+		allowedAsDerivationOutputs(entitiesList.stream().flatMap(List::stream).collect(Collectors.toList()));
+
 		List<String> derivations = new ArrayList<>();
+		Map<String, List<String>> entityInput = new HashMap<>();
 
 		for (int i = 0; i < entitiesList.size(); i++) {
 			List<String> entities = entitiesList.get(i);
 			List<String> inputs = inputsList.get(i);
+			// throw error if any of the IRIs appear in both inputs and entities
+			Set<String> intersection = entities.stream().distinct().filter(inputs::contains).collect(Collectors.toSet());
+			if (!intersection.isEmpty()) {
+				throw new JPSRuntimeException(
+						"Intersection between inputs and outputs for the same derivation markup: " + intersection);
+			}
+			// add pair of output entity and input
+			// this will help to detect same IRI been marked as belongsTo more than one derivation
+			entities.stream().forEach(en -> {
```

Comment:
I don't fullly understand the use of a map here, is this necessary?

Reply:
This will detect the error in situations where the below triples are trying to be added in bulk:
<entity1> <belongsTo> <derivation1>. <derivation1> <isDerivedFrom> <input1>.
<entity1> <belongsTo> <derivation2>. <derivation2> <isDerivedFrom> <input2>.

Reply:
Do you need a map for this? Can't you just have a list and check list.contains(entity)?

Reply:
As discussed, I will leave it as it is now.